src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1776,6 +1777,22 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             }
 
             if let Some((prefix, mpi)) = shortest_uninit_seen {
+                // Check for a reassignment into a uninitialized field of a union (for example,
+                // after a move out). In this case, do not report a error here. There is an
+                // exception, if this is the first assignment into the union (that is, there is
+                // no move out from an earlier location) then this is an attempt at initialization
+                // of the union - we should error in that case.
+                let tcx = this.infcx.tcx;
+                if let ty::TyKind::Adt(def, _) = base.ty(this.mir, tcx).to_ty(tcx).sty {
+                    let moved_before_this = this.move_data.path_map[mpi].iter().any(|moi| {
+                        this.move_data.moves[*moi].source.is_predecessor_of(context.loc, this.mir)
```

Comment:
the `is_predecessor_of` call is O(N) time (where N is the number of locations in the MIR), right? And then you're invoking it here on every path associated with `mpi`... all to compute this boolean which is only used if `def.is_union()`...

I think it would be better (in terms of compiler slowness) for you to delay the computation of `moved_before_this` to only take place *after* you've determined `def.is_union()`.

Reply:
Ah, that's a good point. Thinking about this a little more, the check I've added here is way overkill and checks lots of locations when it doesn't need to. I've updated it so that it does the following:

- If `self` is in the same block as `other`, then check if `self` has a smaller statement index than `other`. If it does, `self` is a predecessor and we can return `true`.
- If `self` is in a different block than other, then all we need to do is check if `self.block` is one of the predecessor blocks of `other`. If it is, return `true`.

This is now linear over the no. of predecessor basic blocks rather than predecessor locations, if I'm not mistaken.

Reply:
great, thanks for your diligence on this!